### PR TITLE
Add uniswap V2 WETH-USDC swap example

### PR DESCRIPTION
**Pull request by pawurb**
Related to https://github.com/bluealloy/revm/issues/1084 . This example  shows common use cases not covered by other examples. 

Screenshot 2024-04-22 at 13 51 58

### CODE REVIEW

File: examples/uniswap_v2_usdc_swap.rs [hunk 1]
```diff
@@ -0,0 +1,357 @@
+use ethers_contract::BaseContract;
```
Comment (onbjerg) [inline examples/uniswap_v2_usdc_swap.rs:1]: please use alloy, ethers is deprecated

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: examples/uniswap_v2_usdc_swap.rs (hunk 1 above)
Comment (pawurb) [reply examples/uniswap_v2_usdc_swap.rs:1]: I have very limited Rust experience, so probably I'm missing something. But is the project currently compatibly with alloy? Adding `alloy-primitives` crate causes dozens of compile errors across the project. I can look into rewriting this example but tips on which crates I should use/avoid would be helpful.

Comment (onbjerg) [reply examples/uniswap_v2_usdc_swap.rs:1]: yes, it is compatible with alloy, it uses alloy internally: https://github.com/bluealloy/revm/blob/1ca3d39f6a9e9778f8eb0fcb74fe529345a531b4/crates/primitives/src/lib.rs

any crate should be fine, but it should use the same version as the crate itself, so currently 0.7: https://github.com/bluealloy/revm/blob/1ca3d39f6a9e9778f8eb0fcb74fe529345a531b4/crates/primitives/Cargo.toml#L20

**Comment by pawurb** [comment]
@onbjerg Should be fixed. I think `ethers_providers` has to stay because of `EthersDB`, or is there some alloy alternative?

**Comment by onbjerg** [comment]
@pawurb you can use `AlloyDB`! 😄

File: examples/uniswap_v2_usdc_swap.rs [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+            output: Output::Call(value),
+            ..
+        } => value,
+        result => panic!("'get_reserves' execution failed: {result:?}"),
+    };
+
+    let (reserve0, reserve1): (u128, u128) = match <(u128, u128, u32)>::abi_decode(&value, false) {
+        Ok((reserve0, reserve1, _)) => (reserve0, reserve1),
+        Err(e) => panic!("'get_reserves' decode failed: {:?}", e),
+    };
```
Comment (onbjerg) [inline examples/uniswap_v2_usdc_swap.rs:257]: you should be able to just use `?` her

Comment (pawurb) [reply examples/uniswap_v2_usdc_swap.rs:260]: I'm getting:

```
the trait `StdError` is not implemented for `alloy_sol_types::Error`, which is required by `std::result::Result<(revm::revm_primitives::alloy_primitives::Uint<256, 4>, revm::revm_primitives::alloy_primitives::Uint<256, 4>), anyhow::Error>: FromResidual<std::result::Result<Infallible, alloy_sol_types::Error>>
```

I can wrap it in `anyhow` error, but it does not look much better than current panic.

Comment (onbjerg) [reply examples/uniswap_v2_usdc_swap.rs:266]: if you want to panic, use `expect` instead

Comment (pawurb) [reply examples/uniswap_v2_usdc_swap.rs]: In the end `?` works with `std` feature enabled.

File: examples/uniswap_v2_usdc_swap.rs [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    let client = ProviderBuilder::new()
+        .on_reqwest_http(
+            "https://mainnet.infura.io/v3/c60b0bb42f8a4c6481ecd229eddaca27"
+                .parse()
+                .unwrap(),
+        )
+        .unwrap();
+    let client = Arc::new(client);
+    let mut alloydb = AlloyDB::new(client, None);
+    let mut cache_db = CacheDB::new(EmptyDB::default());
```
Comment (onbjerg) [inline examples/uniswap_v2_usdc_swap.rs:25]: you can just do `CacheDB::new(AlloyDB::new(client, None))`, then calls to e.g. `db.basic` will automatically first get pulled from the alloy provider, and then stored in the cache db! that avoids you needing to do what you are doing on eg L45, L48 etc

**Review by onbjerg — DISMISSED** [review]

**Review by rakita — APPROVED** [review]

**Comment by rakita** [comment]
pending comment https://github.com/bluealloy/revm/pull/1353#discussion_r1580970513

**Comment by pawurb** [comment]
> pending comment [#1353 (comment)](https://github.com/bluealloy/revm/pull/1353#discussion_r1580970513)

@rakita I don't understand, I think I've already applied suggestions from this comment.

**Comment by rakita** [comment]
> > pending comment [#1353 (comment)](https://github.com/bluealloy/revm/pull/1353#discussion_r1580970513)
> 
> @rakita I don't understand, I think I've already applied suggestions from this comment.

That is true, didn't see that is a outdated comment. lgtm, merging

**Pull request merged**